php/libraries/SinglePointLogin.class.inc

```diff
@@ -308,85 +308,96 @@ class SinglePointLogin
                   GROUP BY UserID";
         $row   = $DB->pselectRow($query, array('username' => $username));
 
-        if (($row['User_count'] ?? 0) == 1) {
-            // validate passsword
-            $oldhash = $row['Password_hash'];
-            if (password_verify($password, $oldhash)) {
-                if (password_needs_rehash($oldhash, PASSWORD_DEFAULT)) {
-                    $user =& User::factory($username);
-                    $user->updatePassword($password);
-                }
-
-                if ($row['Active'] == 'N'
-                    || $this->disabledDueToInactivity($username, 365)
-                ) {
-                    $this->_lastError = "Your account has been deactivated."
-                        . " Please contact your project administrator to"
-                        . " reactivate this account.";
-                    $this->insertFailedDetail(
-                        "user account not active",
-                        $setArray
-                    );
-
-                    return false;
-                }
-
-                // check if the account is longer active
-                $date       = new DateTime();
-                $currentDay = $date->getTimestamp();
-
-                if (($row['active_to'] != null)
-                    && ($currentDay > strtotime($row['active_to']))
-                ) {
-                    $this->_lastError = "Your account has expired."
-                        . " Please contact your project administrator to re-activate"
-                        . " this account.";
-                    return false;
-
-                }
+        if (intval($row['User_count'] ?? 0) !== 1) {
+            /* Display generic failure message to front-end so that an
+             * attacker will have a harder time knowing the difference
+             * between a bad username and a bad password.
+             */
+            $this->_lastError = "Incorrect username or password.";
+            return false;
```

Comment:
if this case returns false, can't it fail by default at the end of this function?

Reply:
It could but it's quicker if it fails out right away instead of going through all these other checks. If someone enters a bad password for a user account it doesn't make sense to go on and check whether that account is expired and tell them that info.